korma/src/commonMain/kotlin/com/soywiz/korma/geom/IRectangle.kt

```diff
@@ -0,0 +1,92 @@
+package com.soywiz.korma.geom
+
+interface IRectangle {
+    val x: Double
+    val y: Double
+    val width: Double
+    val height: Double
+
+    companion object {
+        inline operator fun invoke(
+            x: Double,
+            y: Double,
+            width: Double,
+            height: Double
+        ): IRectangle = Rectangle(x, y, width, height)
+
+        inline operator fun invoke(x: Float, y: Float, width: Float, height: Float): IRectangle =
+            Rectangle(x, y, width, height)
+
+        inline operator fun invoke(x: Int, y: Int, width: Int, height: Int): IRectangle =
+            Rectangle(x, y, width, height)
+
+        // Creates a rectangle from 2 points where the (x,y) is the top left point
+        // with the same width and height as the point.
+        //
+        // Rect XY   Point 2
+        // │        │
+        // ▼        ▼
+        // ┌────────┐
+        // │        │
+        // │        │
+        // └────────┘
+        // ▲
+        // │
+        // Point 1
+        operator fun invoke(point1: IPoint, point2: IPoint): Rectangle {
+            val topLeftX = minOf(point1.x, point2.x)
+            val topLeftY = minOf(point1.y, point2.y)
+            val bottomRightX = maxOf(point1.x, point2.x)
+            val bottomRightY = maxOf(point1.y, point2.y)
+            return Rectangle(topLeftX, topLeftY, bottomRightX - topLeftX, bottomRightY - topLeftY)
+        }
+    }
+}
+
+@Deprecated(
+    "Properties with underscores are deprecated and will be removed soon",
+    ReplaceWith("x"),
+    DeprecationLevel.ERROR
+)
+val IRectangle._x get() = x
+
+@Deprecated(
+    "Properties with underscores are deprecated and will be removed soon",
+    ReplaceWith("y"),
+    DeprecationLevel.ERROR
+)
+val IRectangle._y get() = y
+
+@Deprecated(
+    "Properties with underscores are deprecated and will be removed soon",
+    ReplaceWith("width"),
+    DeprecationLevel.ERROR
+)
+val IRectangle._width get() = width
+
+@Deprecated(
+    "Properties with underscores are deprecated and will be removed soon",
+    ReplaceWith("height"),
+    DeprecationLevel.ERROR
+)
+val IRectangle._height get() = height
+
+val IRectangle.left get() = x
+val IRectangle.top get() = y
+val IRectangle.right get() = x + width
+val IRectangle.bottom get() = y + height
+
+val IRectangle.topLeft get() = Point(left, top)
```

Comment:
This is allocating. Until we have proper inline classes for multiple args, in the APIs we try to use methods with a final argument called out with a default constructor to use them as a holder to avoid allocations. Then there is a PointPool that can have extension methods that avoid allocations. It is a bit painful but a way yo avoid allocation

Reply:
BTW I didnt change this code. Simply moved it to another file. Feel free to submit another PR for that stuff.